MBS-13358 (III): Upgrade Perl version to 5.38.2 for mirrors

yvanzo opened the pull request:
# MBS-13358 (III)

This patch upgrades Perl version to 5.38.2 for the MusicBrainz Server for mirrors.

See commit messages for the details of the implementation which is adapted from the official [docker-perl](https://github.com/Perl/docker-perl/blob/r20240224.0/5.038.002-main%2Cthreaded-bookworm/Dockerfile).

This is the third and last step to implement the above linked ticket.

This pull request is targeting the branch `schema-change-2024-q2` in order to avoid messing up new installations until the schema change is released, and to allow for potential hotfixes to the current release in the meantime if needed.

## Additional changes

Install the Perl module dependencies of MusicBrainz Server using Carton, as in musicbrainz-* Docker images. It prevents breaking the build when a new version of a dependency is released with some incompatibility.

yvanzo committed:
Install local::lib using cpanm instead of apt-get

This method will work even if installing a more recent Perl from source.

Same commit as 28429131224e597f69273eae893c077ccc36529 for normal setup.

yvanzo committed:
Install cpanm from source instead of using apt-get

This method will work even if installing a more recent Perl from source.

Same commit as eabd479464a40d7d9834c894771c20e0c75e1a9 for normal setup.

yvanzo committed:
MBS-13358 (III): Upgrade Perl version to 5.38.2

Our (MetaBrainz) base image is based on Phusion base image which in turn
is based on Ubuntu Jammy that comes with Perl 5.34.0 (using apt).  To
upgrade Perl with the same base image, we have to install from source.

Same commit as 125ec5b894f8614995a0d05f8c4cc03d22cbaab for normal setup.

reosarevok reviewed (commented):
Seems fine AFAICT. Have you tested it and if not how should we test it?

yvanzo commented:
I did set `MUSICBRAINZ_SERVER_VERSION=master`, build image with `--no-cache`, put the mirror website up and resume replication. However I was wrongly using a development setup. Testing it again with a mirror setup failed when building perl without `build-essential` but also when installing `DateTime::Locale` with `cpanm` instead of `carton`. I’m working on a patch which is almost ready but each build takes some time.

yvanzo committed:
Install Perl module dependencies with Carton

Use Carton to install the Perl module dependencies with versions set in
musicbrainz-server/cpanfile.snapshot as in musicbrainz-* Docker images.

Reference: https://metacpan.org/dist/Carton/view/lib/Carton/Doc/FAQ.pod

Setting the environment variable `PERL_CARTON_CPANFILE`
allows running `carton exec` from any current working directory.

Reference: https://github.com/perl-carton/carton/blob/v0.9.63/Changes

A few Perl module dependencies are still installed using cpanm but these
should be moved to the cpanfile in the musicbrainz-server repository.

yvanzo commented:
Now fixed the build and tested against both MBS `v-2024-04-09` and the current `master` branch (requiring Perl 5.38.0).

@reosarevok, please also test it in some way, building, running and using.

yvanzo commented:
Reviewed by @reosarevok on IRC.
